### ISSUE TITLE
Kargo doesn't support simultaneous Dex CLI and Web based SSO login well

### ISSUE THREAD

**Issue by jessesuen** [akuity/kargo#614]
I tried configuring Kargo to handle web and CLI-based SSO login with Dex, but did not have a good time.

Here was my original attempt with dex (take note of `redirectURIs`):

```yaml
connectors:
- config:
    clientID: $CONNECTOR_GITHUB_CLIENT_ID
    clientSecret: $CONNECTOR_GITHUB_CLIENT_SECRET
    orgs:
    - name: akuity
    redirectURI: https://dex.demo.akuity.io/callback
    teamNameField: slug
    useLoginAsID: true
  id: github
  name: GitHub
  type: github
issuer: https://dex.demo.akuity.io
oauth2:
  skipApprovalScreen: true
staticClients:
- id: kargo
  name: Kargo
  public: true
  redirectURIs:
  - https://kargo.demo.akuity.io/login
  - http://localhost
  - http://localhost/auth/callback
storage:
  config:
    inCluster: true
  type: kubernetes
web:
  allowedOrigins:
  - https://kargo.demo.akuity.io
```

The above will work for web login, but fail CLI login with the following error:

```
Bad Request
Unregistered redirect_uri ("http://localhost:57289/auth/callback").
```

The problem is that the port is included as part of the URI, and so the port must be included in dex's config.yaml and exactly match one of the redirectURIs **including the port number**. What makes matters worse, is that `kargo login` will use a random port by default. And so the only way this could work is if the dex admin chose a hard-wired port to register to dex (e.g. 1234). And then that admin must instruct all of their kargo users to run the `kargo login` command with the `--port 1234` flag that they chose to register to dex.

We can do better.

Looking at Dex's `validateRedirectURI` logic: https://github.com/dexidp/dex/blob/70d7a2c7c1bb2646b1a540e49616cbc39622fb83/server/oauth2.go#L632-L636

We see that Dex does allow for randomized ports for CLI login. The catch is that
1. The static client's `redirectURIs` list must be empty
2. The static client must be public (ours already is).
3. The URI being redirected to must be `http://localhost/<subpath>`

So, in other words, this would work:

```yaml
  staticClients:
  - id: kargo
    name: Kargo
    redirectURIs:
    - https://kargo.demo.akuity.io/login
    public: true
  - id: kargo-cli
    name: Kargo CLI
    public: true
```

However, to support the above dex configuration, Kargo must be taught to use *two* separate clientIDs, one for the web and one for the CLI.

To make the UX better, I think we need to support a new, optional `OIDC_CLI_CLIENT_ID` configuration, in addition to the existing `OIDC_CLIENT_ID`. Kargo would understand how to publish both configurations in its public endpoint, for the web and CLI to understand how to use.

Argo CD also uses separate CLI vs. web client with it's bundled dex with success (although it does not use PKCE for web).

**Comment by krancour** [comment]
Dex `staticClients` do not require you to register `redirectURIs` for this exact reason. If you drop `redirectURIs`, you're in business and no changes to Kargo are required.

**Comment by jessesuen** [comment]
No, that won't work either. That results in error:
```
Unregistered redirect_uri ("https://kargo.demo.akuity.io/login").
```

Please take a look at Dex's implementation of validateRedirectURI.
https://github.com/dexidp/dex/blob/70d7a2c7c1bb2646b1a540e49616cbc39622fb83/server/oauth2.go#L624

If redirectUIRs is empty, then the only thing they will accept is localhost (or the named redirect URIS).

**Comment by krancour** [comment]
> If redirectUIRs is empty, then the only thing they will accept is localhost

Ugh. Bummer.

I get it now.
